### PR TITLE
Implement semantic api related visitors for `BLangListConstructorSpreadOpExpr`

### LINKED ISSUE #36073: `ReferenceFinder` crashing with the list constructor spread operator
**Description:**
$subject.

```bal
public function main() {
    int[] x = [1, 2 , 3];
    int[] y = [...x];
}
```

<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
[Error - 11:10:02 AM] Operation 'text/codeAction' failed! {uri: '/home/lochanaj/Downloads/new/test.bal', [5:22]- [5:22], error: 'java.lang.AssertionError, org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor.visit(BLangNodeVisitor.java:661)'}
java.lang.AssertionError
	at org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor.visit(BLangNodeVisitor.java:661)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangListConstructorExpr$BLangListConstructorSpreadOpExpr.accept(BLangListConstructorExpr.java:180)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:224)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:229)
	at io.ballerina.compiler.api.impl.ReferenceFinder.visit(ReferenceFinder.java:895)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangListConstructorExpr.accept(BLangListConstructorExpr.java:57)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:224)
	at io.ballerina.compiler.api.impl.ReferenceFinder.visit(ReferenceFinder.java:340)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:54)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:224)
	at io.ballerina.compiler.api.impl.ReferenceFinder.visit(ReferenceFinder.java:391)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:50)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:224)
	at io.ballerina.compiler.api.impl.ReferenceFinder.visit(ReferenceFinder.java:298)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
	at io.ballerina.compiler.api.impl.ReferenceFinder.find(ReferenceFinder.java:224)
	at io.ballerina.compiler.api.impl.ReferenceFinder.findReferences(ReferenceFinder.java:215)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.getReferences(BallerinaSemanticModel.java:283)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.references(BallerinaSemanticModel.java:229)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.references(BallerinaSemanticModel.java:202)
	at org.ballerinalang.langserver.util.references.ReferencesUtil.lambda$getReferences$0(ReferencesUtil.java:98)
	at java.base/java.util.HashMap$KeySet.forEach(HashMap.java:929)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1085)
	at org.ballerinalang.langserver.util.references.ReferencesUtil.getReferences(ReferencesUtil.java:96)
	at org.ballerinalang.langserver.codeaction.providers.IgnoreUnusedVariableCodeAction.lambda$getDiagBasedCodeActions$2(IgnoreUnusedVariableCodeAction.java:111)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.ballerinalang.langserver.codeaction.providers.IgnoreUnusedVariableCodeAction.lambda$getDiagBasedCodeActions$3(IgnoreUnusedVariableCodeAction.java:111)
	at java.base/java.util.Optional.flatMap(Optional.java:294)
	at org.ballerinalang.langserver.codeaction.providers.IgnoreUnusedVariableCodeAction.getDiagBasedCodeActions(IgnoreUnusedVariableCodeAction.java:110)
	at org.ballerinalang.langserver.codeaction.CodeActionRouter.lambda$getAvailableCodeActions$4(CodeActionRouter.java:122)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.ballerinalang.langserver.codeaction.CodeActionRouter.lambda$getAvailableCodeActions$5(CodeActionRouter.java:113)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
```
</details>

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #36073 

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by lochana-chathura** [comment]
@dulajdilshan @sanjana Do we have semantic api tests for each AST node? Do we need to add test cases with these changes?

**Comment by sanjana** [comment]
> @dulajdilshan @sanjana Do we have semantic api tests for each AST node? Do we need to add test cases with these changes?

@lochana-chathura, 
Need to add a test case for the `references()` in Semantic API for the crashing example mentioned in the related issue.

**Review by lasinicl — APPROVED** [review]

**Comment by pcnfernando** [comment]
Try to handle the uncovered test scenarios for SemanticApi and QueryDesugar

**Review by pcnfernando — APPROVED** [review]

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal [hunk 1]
```diff
@@ -175,3 +175,15 @@ function testFunction() {
 function parseField(string val) returns [string, int] {
     return [val, 10];
 }
+
+int[] x1 = [1, 2 , 3];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal:180]: Superfluous new line. 
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal [hunk 2]
```diff
@@ -175,3 +175,15 @@ function testFunction() {
...
+int[] y1 = [...x1, 4];
+
+function testListConstructorSpreadOp() {
+    int[] x2 = [1, 2 , 3];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal:185]: Ditto
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal [hunk 3]
```diff
@@ -175,3 +175,15 @@ function testFunction() {
...
+    int[] y2 = [...x2, 4];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/find-all-ref/find_var_ref_in_exprs.bal:187]: Ditto
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal [hunk 4]
```diff
@@ -196,3 +196,15 @@ function fn() {
 function add(int a, int b) returns int {
     return 1;
 }
+
+int[] x1 = [1, 2 , 3];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal:201]: Ditto
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal [hunk 5]
```diff
@@ -196,3 +196,15 @@ function fn() {
...
+int[] y1 = [...x1, 4];
+
+function testListConstructorSpreadOp() {
+    int[] x2 = [1, 2 , 3];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal:206]: Ditto
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal [hunk 6]
```diff
@@ -196,3 +196,15 @@ function fn() {
...
+    int[] y2 = [...x2, 4];
+
```
Comment (sanjana) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/symbol_at_cursor_basic_test.bal:208]: Ditto
```suggestion
```

**Review by sanjana — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source5.bal [hunk 7]
```diff
@@ -0,0 +1,4 @@
+function testListConstructorSpreadOp() {
+    int[] x = [1, 2];
+    int[] y = [...];
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source5.bal:3]: We have to fix the sorting of the completions at the cursor position after `...`. Further, we may have to add a spread op completion item as well (Just like spread field completion item for mapping constructors). Will create a separate issue for this.

**Review by malinthar — APPROVED** [review]
LGTM!

**Review by sanjana — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source5.bal (hunk 7 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source5.bal:3]: https://github.com/ballerina-platform/ballerina-lang/issues/35507 tracks the changes to be made

**Pull request merged**